modules/base/rendering/screenspaceinsetblackout.cpp

```diff
@@ -0,0 +1,951 @@
+
+        // Incoming vertex data should be top -> right -> bottom -> left (clockwise)
+        // Data to render needs to be counter-clockwise for correct winding
+        // Also adds an extra point at the end for Triangle Fan
+        _vertexBufferData.push_back(glm::vec2(0.0, 0.0));
+        copyToVertexBuffer(splineTop);
+        copyToVertexBuffer(splineLeft);
+        copyToVertexBuffer(splineBottom);
+        copyToVertexBuffer(splineRight);
+        _vertexBufferData.push_back(splineTop.back());
+    }
+}
+
+void ScreenSpaceInsetBlackout::copyToPointsVector(
+                            const std::vector<std::unique_ptr<Spline::Point>>& points,
+                            std::vector<glm::vec2>& vertexData)
+{
+    for (int i = 0; i < points.size(); ++i) {
+        vertexData.push_back(*points[i]->prop);
+    }
+}
+
+void ScreenSpaceInsetBlackout::copyToVertexBuffer(const std::vector<glm::vec2>& v) {
+    _vertexBufferData.insert(
+        _vertexBufferData.end(),
+        v.rbegin(),
+        v.rend()
+    );
+}
+
+std::vector<glm::vec2> ScreenSpaceInsetBlackout::sampleSpline(
+                                            const std::vector<glm::vec2>& controlPoints)
+{
+    std::vector<glm::vec2> outSplineData;
+    const int numberOfSegments = static_cast<int>(controlPoints.size() - 3);
+    const float stepSize = 1.f / SUBDIVIDES;
+    for (int i = 0; i < numberOfSegments; ++i) {
+        for (int s = 0; s < SUBDIVIDES; ++s) {
+            float tValue = stepSize * s;
+            glm::vec2 splinePosition = calculateCatmullRom(
+                *(controlPoints.begin() + i + 0),
+                *(controlPoints.begin() + i + 1),
+                *(controlPoints.begin() + i + 2),
+                *(controlPoints.begin() + i + 3),
+                tValue
+            );
+            outSplineData.push_back(splinePosition);
+        }
+    }
+    return outSplineData;
+}
+
+glm::vec2 ScreenSpaceInsetBlackout::calculateCatmullRom(const glm::vec2& p0,
+                                                        const glm::vec2& p1,
+                                                        const glm::vec2& p2,
+                                                        const glm::vec2& p3,
+                                                        const float t)
+{
+    glm::vec2 newPoint;
+    const float t3 = powf(t, 3.f);
+    const float t2 = powf(t, 2.f);
+    const float alpha = 0.5f;
+
+    const float p0t = ((-t) + (2.f * t2) - t3);
+    const float p1t = (2.f - (5.f * t2) + (3.f * t3));
+    const float p2t = (t + (4.f * t2) - (3.f * t3));
+    const float p3t = (-t2 + t3);
+
+    const glm::vec2 np0 = p0 * p0t;
+    const glm::vec2 np1 = p1 * p1t;
+    const glm::vec2 np2 = p2 * p2t;
+    const glm::vec2 np3 = p3 * p3t;
+
+    newPoint = alpha * (np0 + np1 + np2 + np3);
+
+    return newPoint;
+}
+
+void ScreenSpaceInsetBlackout::generateTexture() {
+    //Generate new vertex array data
+    generateVertexArray();
```

Comment:
A bit more description would be useful;  it is not clear to me why a call to generate texture would need to regenerate the VAO aswell

Reply:
`generateTexture` should be renamed to `renderTexture`